In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split 
import time
import numpy as np
from sklearn.preprocessing import StandardScaler
from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import chi2
from sklearn.feature_selection import RFE
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report
import pickle
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings("ignore")

In [2]:
def split_data(indep_X,dep_Y):
    X_train,X_test,y_train,y_test=train_test_split(indep_X,dep_Y,test_size=0.25,random_state=0,stratify=dep_Y)
    return X_train,X_test,y_train,y_test

In [3]:
def selectkbest(X_train,y_train,X_test,n):
    test=SelectKBest(score_func=chi2,k=n)
    fit1=test.fit(X_train,y_train)
    X_train_selected=fit1.transform(X_train)
    X_test_selected=fit1.transform(X_test)
    return X_train_selected,X_test_selected

In [4]:
def scale_data(X_train,X_test):
    sc=StandardScaler()
    X_train=sc.fit_transform(X_train)
    X_test=sc.transform(X_test)
    return X_train,X_test

In [5]:
def counterfeit_prediction(classifier,X_test,y_test):
    y_pred=classifier.predict(X_test)
    cft=confusion_matrix(y_test,y_pred)
    Accuracy=accuracy_score(y_test,y_pred)
    report=classification_report(y_test,y_pred)
    return classifier,Accuracy,report,X_test,y_test,cft

In [6]:
def logistic(X_train,y_train,X_test,y_test):
    classifier=LogisticRegression(random_state=0)
    classifier.fit(X_train,y_train)
    classifier,Accuracy,report,X_test,y_test,cft=counterfeit_prediction(classifier,X_test,y_test)
    return classifier,Accuracy,report,X_test,y_test,cft

In [7]:
def svm_linear(X_train,y_train,X_test,y_test):
    from sklearn.svm import SVC
    classifier=SVC(kernel='linear',random_state=0)
    classifier.fit(X_train,y_train)
    classifier,Accuracy,report,X_test,y_test,cft=counterfeit_prediction(classifier,X_test,y_test)
    return classifier,Accuracy,report,X_test,y_test,cft

In [8]:
def svm_NL(X_train,y_train,X_test,y_test):
    from sklearn.svm import SVC
    classifier=SVC(kernel='rbf',random_state=0)
    classifier.fit(X_train,y_train)
    classifier,Accuracy,report,X_test,y_test,cft=counterfeit_prediction(classifier,X_test,y_test)
    return classifier,Accuracy,report,X_test,y_test,cft

In [9]:
def Navie(X_train,y_train,X_test,y_test):
    from sklearn.naive_bayes import GaussianNB
    classifier=GaussianNB()
    classifier.fit(X_train,y_train)
    classifier,Accuracy,report,X_test,y_test,cft=counterfeit_prediction(classifier,X_test,y_test)
    return classifier,Accuracy,report,X_test,y_test,cft

In [10]:
def knn(X_train,y_train,X_test,y_test):
    from sklearn.neighbors import KNeighborsClassifier
    classifier=KNeighborsClassifier(n_neighbors=5,metric='minkowski',p=2)
    classifier.fit(X_train,y_train)
    classifier,Accuracy,report,X_test,y_test,cft=counterfeit_prediction(classifier,X_test,y_test)
    return classifier,Accuracy,report,X_test,y_test,cft

In [11]:
def Decision(X_train,y_train,X_test,y_test):
    from sklearn.tree import DecisionTreeClassifier
    classifier=DecisionTreeClassifier(criterion='entropy',random_state=0)
    classifier.fit(X_train,y_train)
    classifier,Accuracy,report,X_test,y_test,cft=counterfeit_prediction(classifier,X_test,y_test)
    return classifier,Accuracy,report,X_test,y_test,cft

In [12]:
def random(X_train,y_train,X_test,y_test):
    from sklearn.ensemble import RandomForestClassifier
    classifier=RandomForestClassifier(n_estimators=10,criterion='entropy',random_state=0)
    classifier.fit(X_train,y_train)
    classifier,Accuracy,report,X_test,y_test,cft=counterfeit_prediction(classifier,X_test,y_test)
    return classifier,Accuracy,report,X_test,y_test,cft

In [13]:
def selectk_Classification(acclog,accsvml,accsvmnl,accknn,accnav,accdes,accrf):
    dataframe=pd.DataFrame(index=['ChiSquare'],columns=['Logistic','SVMl','SVMnl','KNN','Navie','Decision','Random'])
    for number,idex in enumerate(dataframe.index):
        dataframe['Logistic'][idex]=acclog[number]
        dataframe['SVMl'][idex]=accsvml[number]
        dataframe['SVMnl'][idex]=accsvmnl[number]
        dataframe['KNN'][idex]=accknn[number]
        dataframe['Navie'][idex]=accnav[number]
        dataframe['Decision'][idex]=accdes[number]
        dataframe['Random'][idex]=accrf[number]
    return dataframe

In [14]:
cft_pdt_data = pd.read_csv("counterfeit_products_balanced_preprocessed_v2.csv",index_col=None)
cft_pdt_data

,price,seller_rating,seller_reviews,product_images,description_length,shipping_time_days,spelling_errors,domain_age_days,contact_info_complete,return_policy_clear,...,ip_location_mismatch,seller_country,shipping_origin,brand_suspicious,category_Cosmetics,category_Electronics,category_Fashion,category_Luxury Goods,category_Pharmaceuticals,is_counterfeit
0,277.200000,3.500000,890,4,128.000000,21.000000,0,1440,0,1,...,0,1,6,0,0,0,1,0,0,0
1,89.180000,3.978268,1859,3,224.000000,31.000000,4,206,0,0,...,1,5,5,0,0,1,0,0,0,1
2,168.400000,2.700000,1096,2,98.000000,10.000000,0,1069,1,1,...,0,7,1,1,0,0,1,0,0,1
3,130.980000,3.100000,1296,3,255.000000,22.000000,0,514,1,1,...,0,7,6,0,0,0,0,0,0,0
4,5.000000,2.600000,0,2,20.000000,32.000000,5,5,0,0,...,0,8,9,0,0,0,0,0,1,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
5899,227.623841,3.037227,86,1,510.672921,9.360590,1,1864,1,0,...,0,2,2,1,0,0,0,0,0,1
5900,144.777008,3.422818,734,1,138.623748,32.850277,0,1046,0,0,...,0,7,4,0,0,0,0,0,0,1
5901,176.776416,3.822738,1988,1,424.930209,14.726227,0,308,0,1,...,0,7,3,1,0,0,1,0,0,1
5902,127.274290,3.500000,122,1,273.755228,22.026568,0,5,1,0,...,0,2,3,0,0,0,0,0,0,1


In [15]:
indep_X=cft_pdt_data.drop('is_counterfeit',axis=1)
dep_Y=cft_pdt_data['is_counterfeit']

In [37]:
X_train,X_test,y_train,y_test=split_data(indep_X,dep_Y)
X_train,X_test=selectkbest(X_train,y_train,X_test,20)
X_train,X_test=scale_data(X_train,X_test)
X_train.shape,X_test.shape

((4428, 20), (1476, 20))

In [38]:
acclog=[]
accsvml=[]
accsvmnl=[]
accknn=[]
accnav=[]
accdes=[]
accrf=[]


classifier,Accuracy,report,X_test,y_test,cft=logistic(X_train,y_train,X_test,y_test)
acclog.append(Accuracy)

classifier,Accuracy,report,X_test,y_test,cft=svm_linear(X_train,y_train,X_test,y_test)
accsvml.append(Accuracy)

classifier,Accuracy,report,X_test,y_test,cft=svm_NL(X_train,y_train,X_test,y_test)
accsvmnl.append(Accuracy)

classifier,Accuracy,report,X_test,y_test,cft=knn(X_train,y_train,X_test,y_test)
accknn.append(Accuracy)

classifier,Accuracy,report,X_test,y_test,cft=Navie(X_train,y_train,X_test,y_test)
accnav.append(Accuracy)

classifier,Accuracy,report,X_test,y_test,cft=Decision(X_train,y_train,X_test,y_test)
accdes.append(Accuracy)

classifier,Accuracy,report,X_test,y_test,cft=random(X_train,y_train,X_test,y_test)
accrf.append(Accuracy)

result=selectk_Classification(acclog,accsvml,accsvmnl,accknn,accnav,accdes,accrf)

In [21]:
#5
result

,Logistic,SVMl,SVMnl,KNN,Navie,Decision,Random
ChiSquare,0.630759,0.631436,0.636179,0.630081,0.625339,0.594851,0.629404


In [24]:
#7
result

,Logistic,SVMl,SVMnl,KNN,Navie,Decision,Random
ChiSquare,0.640921,0.647019,0.653794,0.611111,0.643631,0.592818,0.633469


In [27]:
#10
result

,Logistic,SVMl,SVMnl,KNN,Navie,Decision,Random
ChiSquare,0.705962,0.705285,0.707995,0.690379,0.708672,0.672087,0.715447


In [30]:
#12
result

,Logistic,SVMl,SVMnl,KNN,Navie,Decision,Random
ChiSquare,0.698509,0.70122,0.710705,0.695122,0.710027,0.686992,0.70935


In [33]:
#15
result

,Logistic,SVMl,SVMnl,KNN,Navie,Decision,Random
ChiSquare,0.712737,0.70935,0.73103,0.682927,0.710027,0.682927,0.722222


In [36]:
#18
result

,Logistic,SVMl,SVMnl,KNN,Navie,Decision,Random
ChiSquare,0.72019,0.719512,0.730352,0.696477,0.72561,0.662602,0.73645


In [39]:
#20
result

,Logistic,SVMl,SVMnl,KNN,Navie,Decision,Random
ChiSquare,0.732385,0.73103,0.743225,0.686992,0.723577,0.657859,0.72561


#### Taking K=10 and creating model with Random Forest as it has the best score of 71%